via_verifier/lib/via_musig2/src/fee.rs

```diff
@@ -0,0 +1,202 @@
+use std::vec;
+
+use bitcoin::{Amount, TxOut};
+
+pub trait FeeStrategy: Send + Sync {
+    fn estimate_fee(
+        &self,
+        input_count: u32,
+        output_count: u32,
+        fee_rate: u64,
+    ) -> anyhow::Result<Amount> {
+        // version + locktime
+        let base_size = 10_u64;
+        // approximate size per input
+        let input_size = 148_u64 * u64::from(input_count);
+        // approximate size per output +2 (+1 for OP_RETURN, +1 for potential change)
+        let output_size = 34_u64 * u64::from(output_count + 2);
+
+        let total_size = base_size + input_size + output_size;
+        let fee = fee_rate * total_size;
+
+        let remainder = fee % std::cmp::max(output_count, 1) as u64;
+
+        Ok(Amount::from_sat(fee + remainder))
+    }
+
+    fn apply_fee_to_inputs(
+        &self,
+        outputs: Vec<TxOut>,
+        input_count: u32,
+        fee_rate: u64,
+    ) -> anyhow::Result<(Vec<TxOut>, Amount, Amount)>;
+}
+
+pub struct WithdrawalFeeStrategy {}
+
+impl WithdrawalFeeStrategy {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl FeeStrategy for WithdrawalFeeStrategy {
+    fn apply_fee_to_inputs(
+        &self,
+        mut outputs: Vec<TxOut>,
+        input_count: u32,
+        fee_rate: u64,
+    ) -> anyhow::Result<(Vec<TxOut>, Amount, Amount)> {
+        loop {
+            let fee = self.estimate_fee(input_count, outputs.len() as u32, fee_rate)?;
+            if outputs.is_empty() {
+                return Ok((vec![], fee, Amount::ZERO));
+            }
+
+            let fee_per_user = Amount::from_sat(fee.to_sat() / outputs.len() as u64);
+            let mut new_outputs = vec![];
+            let mut new_outputs_with_fee = vec![];
+            let mut total_value_needed = Amount::ZERO;
+
+            for output in &outputs {
+                if output.value < fee_per_user {
+                    continue; // This output can't pay the fee
+                }
+
+                let value = output.value - fee_per_user;
+                new_outputs_with_fee.push(TxOut {
+                    script_pubkey: output.script_pubkey.clone(),
+                    value,
+                });
+                new_outputs.push(TxOut {
+                    script_pubkey: output.script_pubkey.clone(),
+                    value: output.value,
+                });
+                total_value_needed += value
+            }
+
+            if new_outputs.len() != outputs.len() {
+                // Output set has changed, need to re-estimate fee
+                outputs = new_outputs;
+                continue;
+            }
```

Comment:
One suggestion, maybe we can optimize this logic a bit and not create `new_outputs` vector every time, but compare `if new_outputs_with_fee.len() != outputs.len()` and then just recreated the vector that we need to pass as the new `outputs` by adding back the `fee_per_user` to the value.